### DISCUSSION TITLE
Security issue with Github access tokens, how to do it properly?

### DISCUSSION THREAD

**Discussion by AnomalRoil** [goreleaser/goreleaser#2028] [Q&A] [answered]
Hello, while working on GoReleaser integration into [Gopass](https://github.com/gopasspw/gopass/) CI/DI with @dominikschulz, we stumbled upon this security gap and would love some extra input.

## The issue
According to [GoReleaser's doc](https://goreleaser.com/ci/actions/#limitation), one is supposed to use a custom Access Token to publish taps to Homebrew:
> If you need to push the homebrew tap to another repository, you must therefore create a custom [Personal Access Token](https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/) with repo permissions and [add it as a secret in the repository](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets). 

However it does not say how these tokens are supposed to be generated, linking to the generic Github documentation.

Worse, in [another doc entry](https://goreleaser.com/environment/#api-tokens) it even has a direct link to [one's own settings](https://github.com/settings/tokens/new), clearly encouraging people to create a **Personal** Access Token with `repo ` rights.

Notice that using Personal Access Token in a CI/DI is a **really bad idea** because these tokens are granting access to **all the repo** of the user who created it. This is not something individual contributors in an organization usually want to do as it would expose their repo to any "malicious" CI/DI job.  These tokens are meant to be Personal, as their name implies.

Since the current documentation is encouraging people to do it without any warning or caveats, I've opened an issue about it: #2026

Now, in #1643 it says:
> This makes it inflexible for organisations with good security hygiene where Homebrew tap and Scoop bucket may be each managed centrally and tokens issued separately with just access to each individual repository.

But it appears that Github [does not have the ability to issue token with restricted access to a single repository](https://github.community/t/limiting-scope-of-a-pat-to-a-single-repository/3129).

## The possible workarounds
A very hackish workaround would be to use Installation tokens from Github Apps, as [these can be restricted to certain repo only](https://docs.github.com/en/developers/apps/authenticating-with-github-apps#authenticating-as-an-installation).

Another workaround is to create a [Machine User, as these are authorized by Github](https://docs.github.com/en/developers/overview/managing-deploy-keys#machine-users) and use a Personal Access Token from that Machine User (with restricted rights to the required repos).

Ideally we could also use Deploy Keys directly from goreleaser, as deploy keys can be provisioned per repo. I've opened a feature request on the topic: #2027

## Discussion

So, what is your take on the topic?

How do you securely deploy tokens that do not enable access from the CI/DI to all the repos of a user? 

@radeksimko I see you authored the above issue and related PR introducing dedicated Access Tokens for Homebrew and Scoop, how are you supposed to generate these tokens?

**Comment by radeksimko** [comment, marked as answer]
Hi @AnomalRoil 
You are right that the current situation isn't ideal.

> A very hackish workaround would be to use Installation tokens from Github Apps, as these can be restricted to certain repo only.

That does sound very hackish.

> Another workaround is to create a Machine User, as these are authorized by Github and use a Personal Access Token from that Machine User (with restricted rights to the required repos).

Yes, that is pretty much what we do. It's probably the best solution within the existing constraints. Strictly speaking though one should probably have a machine user dedicated to each repository, or at least different ones based on perceived risk profiles, because the token itself just inherits whatever access privileges the user has, as you already mentioned.

> Ideally we could also use Deploy Keys directly from goreleaser, as deploy keys can be provisioned per repo. I've opened a feature request on the topic: #2027

From what I understand this would require goreleaser to clone the repo and in general speak SSH, which it does not today and introducing that would not be as trivial.

We have discussed this option under https://github.com/goreleaser/goreleaser/issues/1774

--- 

I think the ideal solution would involve GitHub introducing one of the following:

 - more granular scoping for tokens, at the very least repo scoping. That would at least make it safer to reuse the same machine user, but maintain different PATs per repositories.
 - API-compatible deploy keys, similar to how the existing deploy keys work, just with the ability to use them with the standard REST v2/v3 API

**Reply by caarlos0** [reply]
Was typing basically the same thing as @radeksimko .

I think we can maybe improve the docs on this a bit? Although I don't quite see the issue with linking to current's user settings and etc... 

Also fair to point out that most of the docs were written with "a user trying to release its own stuff" in mind - which is usually the simpler case and also the less-experienced people who need more help. 

Orgs can have different security restrictions, and I don't think its our responsibility to cover those while making our docs more complicated for new users, although maybe its fair to add heads up here and there about basic sec stuff.
